src/librustdoc/fold.rs

```diff
@@ -50,7 +55,40 @@ pub trait DocFolder: Sized {
                 TraitItem(i)
             }
             ImplItem(mut i) => {
-                i.items = i.items.into_iter().filter_map(|x| self.fold_item(x)).collect();
+                let replacer_opt = if let Type::ResolvedPath { path, .. } = &i.for_ {
+                    path.segments.first().map(|segment| format!("${{1}}{}::${{2}}", segment.name))
+                } else {
+                    None
+                };
+                i.items = i
+                    .items
+                    .into_iter()
+                    .filter_map(|mut x| {
+                        if let (Some(_), Some(replacer)) = (&x.name, &replacer_opt) {
+                            // This whole if block allows to use `Self::<method_name>` in method's doc
+                            // by replacing
+                            // "Self::<method_name>" inside documentation
+                            // with
+                            // <inner.for_.path.segements>::<method>
+                            x.attrs.doc_strings = x
+                                .attrs
+                                .doc_strings
+                                .into_iter()
+                                .map(|doc_frag| {
+                                    if let DocFragment::SugaredDoc(line, span, text) = doc_frag {
+                                        let new_doc_line = SELF_MATCHER
+                                            .replace_all(text.as_ref(), replacer.as_str())
```

Comment:
I'm not sure to understand why we need a regex replacement here... Isn't it possible to just remove the first 6 characters if it starts with "Self::" and prepend with the new text instead? I think this is completely overkill.

Reply:
the 6 characters should work in most cases fine.
However, there is one case when it fails, it's `Self` in name, e.g.
```
/// Use MySelf::new to ..
struct MySelf 
```
1. for regex it will ends up like this
```
/// Use MySelf::new to ..
struct MySelf 
```
2. for the 6 replacement ..
```
/// Use MyMySelf::new to ..
struct MySelf 
```
these test cases cover that
https://github.com/rust-lang/rust/pull/71289/files#diff-aefbb1cc61440df9316be7cd6b0f7c2eR66-R82 


On the other hand, i can imagine that using `Self` in the struct name is not common case, so it could be skipped as well. I feel it should be covered so i did.

Reply:
The important part here was:

> if it starts with "Self::"

;)

Reply:
Does it mean it's ok now, are should i change it? :)
If we're changing it, i need more details or ready-to-apply solution ..
Also I'm open to discuss it live somewhere..
I'm on Rust's
1. Discord 
2. Riot

Reply:
you can ping in #rustdoc channel on official discord server

Reply:
When adding new dependencies, one should have good arguments to support it. You're adding both `regex` and `lazy_static`/`lazy_cell` for one thing. And I'm still not sure to understand *why* it's needed. From what I understand, if the string starts with `Self::`, you want to replace it. So why would a regex be needed there?

Reply:
Did you see https://github.com/rust-lang/rust/pull/71289#discussion_r411214750
?

Reply:
No it doesn't. If it doesn't start with `Self::`, then you don't apply this. Otherwise I'm missing something but can't see what...

Reply:
We can't speak of "start with" because we need to replace substrings. 
E.g.

```
[my link](Self::Struct)
```

Maybe, please provide your code for the replacement part and i'll use it?

Reply:
On the other hand all possible values are
```
[link](Self::Struct)
[link]: Self::Iterator
[link]:\tSelf::Iterator
[link]:Self::Iterator
```
if my correct.

They could be handled with a ifs, splits and statswith as you suggest.

Reply:
The string is really all `[link](Self::Struct)`? If so I'm very surprised... In this case, I thought it was only "Self::Struct". Can you print the actual string at this point please?

Reply:
> Can you print the actual string at this point please?

Which string? :) There can be many string, eg.

```
[link](Self::Struct)
[link]: Self::Iterator
[link]:\tSelf::Iterator
[link]:Self::Iterator
```

Reply:
Also, see
https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md#detailed-design